azurecosmos/src/main/java/site/ycsb/db/AzureCosmosClient.java

```diff
@@ -187,218 +245,251 @@ private int getIntProperty(String propertyName, int defaultValue) {
+   *
+   * @param table       The name of the table
+   * @param startkey    The record key of the first record to read.
+   * @param recordcount The number of records to read
+   * @param fields      The list of fields to read, or null for all of them
+   * @param result      A Vector of HashMaps, where each HashMap is a set
+   *                    field/value pairs for one record
+   * @return Zero on success, a non-zero error code on error
+   */
   @Override
   public Status scan(String table, String startkey, int recordcount, Set<String> fields,
-                     Vector<HashMap<String, ByteIterator>> result) {
-    List<Document> documents;
-    FeedResponse<Document> feedResponse = null;
+      Vector<HashMap<String, ByteIterator>> result) {
     try {
-      FeedOptions feedOptions = new FeedOptions();
-      feedOptions.setEnableCrossPartitionQuery(true);
-      feedOptions.setMaxDegreeOfParallelism(this.maxDegreeOfParallelismForQuery);
-      feedResponse = AzureCosmosClient.client.queryDocuments(getDocumentCollectionLink(this.databaseName, table),
-            new SqlQuerySpec("SELECT TOP @recordcount * FROM root r WHERE r.id >= @startkey",
-                    new SqlParameterCollection(new SqlParameter("@recordcount", recordcount),
-                                               new SqlParameter("@startkey", startkey))),
-                    feedOptions);
-      documents = feedResponse.getQueryIterable().toList();
-    } catch (Exception e) {
-      if (!this.includeExceptionStackInLog) {
-        e = null;
+      CosmosQueryRequestOptions queryOptions = new CosmosQueryRequestOptions();
+      queryOptions.setMaxDegreeOfParallelism(AzureCosmosClient.maxDegreeOfParallelism);
+      queryOptions.setMaxBufferedItemCount(AzureCosmosClient.maxBufferedItemCount);
+
+      CosmosContainer container = database.getContainer(table);
+
+      List<SqlParameter> paramList = new ArrayList<>();
+      paramList.add(new SqlParameter("@startkey", startkey));
+
+      SqlQuerySpec querySpec = new SqlQuerySpec(
+          this.createSelectTop(fields, recordcount) + " FROM root r WHERE r.id >= @startkey", paramList);
+      CosmosPagedIterable<ObjectNode> pagedIterable = container.queryItems(querySpec, queryOptions, ObjectNode.class);
+      Iterator<FeedResponse<ObjectNode>> pageIterator = pagedIterable
+          .iterableByPage(AzureCosmosClient.preferredPageSize).iterator();
+      while (pageIterator.hasNext()) {
+        List<ObjectNode> pageDocs = pageIterator.next().getResults();
+        for (ObjectNode doc : pageDocs) {
+          Map<String, String> stringResults = new HashMap<>();
+          Iterator<Map.Entry<String, JsonNode>> nodeIterator = doc.fields();
+          while (nodeIterator.hasNext()) {
+            Entry<String, JsonNode> pair = nodeIterator.next();
+            stringResults.put(pair.getKey().toString(), pair.getValue().toString());
+          }
+          HashMap<String, ByteIterator> byteResults = new HashMap<>();
```

Comment:
ditto
```suggestion
          Map<String, String> stringResults = new HashMap<>(doc.size());
```

Reply:
Keeping as HashMap since the result parameter is expecting a Vector of HashMaps.

Reply:
the comment is around pre-allocating the memory rather than the type
```new HashMap<>(doc.size());```